saleor/order/utils.py

```diff
@@ -163,37 +163,54 @@ def _calculate_quantity_including_returns(order):
     return total_quantity, quantity_fulfilled, quantity_returned
 
 
+def refresh_order_status(order: Order):
+    """Refresh order status based on the most recent data.
+
+    This function recalculates the order status using the most up-to-date information
+    about fulfillments, returns, and replacements. It should always be called within
+    a transaction and with the order locked to ensure data consistency and prevent race conditions.
+
+    Returns
+        bool: True if the order status was changed, False otherwise.
+
+    """
+    old_status = order.status
+    # Calculate the quantities for the most recent data
+    (
+        total_quantity,
+        quantity_fulfilled,
+        quantity_returned,
+    ) = _calculate_quantity_including_returns(order)
+
+    # check if order contains any fulfillments that awaiting approval
+    awaiting_approval = order.fulfillments.filter(
+        status=FulfillmentStatus.WAITING_FOR_APPROVAL
+    ).exists()
+
+    order.status = determine_order_status(
+        order,
+        total_quantity,
+        quantity_fulfilled,
+        quantity_returned,
+        awaiting_approval,
+    )
+    return old_status != order.status
+
+
 def update_order_status(order: Order):
     """Update order status depending on fulfillments."""
     with transaction.atomic():
         # Add a transaction block to ensure that the order status won't be overridden by
         # another process.
         locked_order = Order.objects.select_for_update().get(pk=order.pk)
-        # Calculate the quantities for the most recent data
-        (
-            total_quantity,
-            quantity_fulfilled,
-            quantity_returned,
-        ) = _calculate_quantity_including_returns(locked_order)
-
-        # check if order contains any fulfillments that awaiting approval
-        awaiting_approval = locked_order.fulfillments.filter(
-            status=FulfillmentStatus.WAITING_FOR_APPROVAL
-        ).exists()
-
-        status = determine_order_status(
-            order,
-            total_quantity,
-            quantity_fulfilled,
-            quantity_returned,
-            awaiting_approval,
-        )
+
+        status_updated = refresh_order_status(locked_order)
 
         # we would like to update the status for the order provided as the argument
         # to ensure that the reference order has up to date status
-        if status != order.status:
-            order.status = status
-            order.save(update_fields=["status", "updated_at"])
+        if status_updated:
+            order.status = locked_order.status
+            locked_order.save(update_fields=["status", "updated_at"])
```

Comment:
Why you changed here to `locked_order`, in some places we are updating the provided order and then operate on the `order` not `locked_order` and some mutations might return invalid status because of that

Reply:
I added a comment to ensure that `order` and `locked_order` have the same status, to ensure that the mutation response contains proper data. 
`refresh_order_status`  should use only one order object. This function is required in transaction mutations to ensure that all prices and statuses for the order are calculated for the same locked order.